#! /bin/sh -e

# DP: Fix strncpy

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
     -patch) patch -d "$2" -f --no-backup-if-mismatch -p0 < $0;;
     -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p0 < $0;;
     *)
         echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
         exit 1
esac
exit 0
			    
--- sysdeps/ia64/strncpy.S~	Tue Apr  2 18:38:41 2002
+++ sysdeps/ia64/strncpy.S	Tue Apr  2 21:27:06 2002
@@ -205,6 +206,7 @@
 	st1	[in0] = c, 1		// *dest++ = c
 (p6)	cmp.ne	p6, p0 = c, r0
 	br.cloop.dptk .l8
+	;;
 	mov 	ar.lc = saved_lc	// restore the loop counter
 	mov	pr = saved_pr, -1	// restore the predicate registers
 	br.ret.sptk.many b0
@@ -213,11 +215,11 @@
 	ld8	r[0] = [tmp]
 	br.cond.sptk .back2
 .recovery3:
-	add	tmp = -(MEMLAT + 1) * 8, src ;;
+	add	tmp = -MEMLAT * 8, src ;;
 	ld8	r[MEMLAT] = [tmp]
 	br.cond.sptk .back3
 .recovery4:
-	add	tmp = -(MEMLAT + 1) * 8, src ;;
-	ld8	r[MEMLAT] = [tmp]
+	add	tmp = -(MEMLAT - 1) * 8, src ;;
+	ld8	r[MEMLAT-1] = [tmp]
 	br.cond.sptk .back4
 	.end
